## Thiên vị - Phương sai (Bias - Variance Dilemma) - Tại sao chúng ta sử dụng kiểm định?

Làm thế nào chúng ta có thể chắc chắn rằng thuật toán của mình thực sự đại diện cho **sự thật ngầm** $F(X)$ chứ không phải là đang **học thuộc lòng** bộ dữ liệu $y$? Chúng ta biết rằng việc *thực sự* biết $F(X)$ là điều không thể. Trong hầu hết các trường hợp, chúng ta cũng sẽ không học thuộc lòng hoàn toàn bộ dữ liệu. Hàm ước lượng $\\hat{F(X))}$ được tạo ra sẽ là một cái gì đó ở giữa.

<img src="img/biasVarianceCurve.jpg" width="70%">

Nguồn: https://www.quora.com/What-is-the-best-way-to-explain-the-bias-variance-trade-off-in-layman%E2%80%99s-terms

Vấn đề này được gọi là **Đánh đổi Thiên vị-Phương sai (Bias-Variance tradeoff)**. Lý thuyết đằng sau nó đã được thiết lập rõ ràng và mô tả chính xác bằng toán học trong các tài liệu. Ở đây chúng ta sẽ tập trung vào trực giác.

Càng cho hàm ước lượng của chúng ta nhiều **sự tự do** (độ đàn hồi/phương sai cao hơn), nó càng dễ dàng và tốt hơn trong việc khớp với dữ liệu huấn luyện. Do đó, nó sẽ đạt được **lỗi thấp hơn** (thiên vị) trên tập huấn luyện. Đồng thời, khả năng biến động càng cao thì **nguy cơ quá khớp (overfitting)** càng lớn. Mặc dù về mặt lý thuyết, chúng ta có thể giảm giá trị kỳ vọng của lỗi (thiên vị), nhưng điều này thường đi kèm với việc **tăng phương sai của lỗi**. Chúng ta có thể rơi vào tình huống mà khi dự đoán trên dữ liệu mới, lỗi rất cao do phương sai (mặc dù trung bình lỗi có thể bằng không).

Để hiểu rõ hơn, hãy xem hình ảnh này:
<img src="img/biasVarianceTarget.jpg" width="30%">

Nguồn: https://www.quora.com/What-is-the-best-way-to-explain-the-bias-variance-trade-off-in-layman%E2%80%99s-terms

Trong thực tế, phương sai cao hơn thường có nghĩa là **lỗi cao hơn** (trên dữ liệu mới). Vấn đề này được thể hiện rõ ràng trong hình minh họa sau.
<img src="img/biasVarianceValid.png" width="50%">
Nguồn: Elements of Statistical learning

Đường **màu đỏ** biểu thị lỗi trên dữ liệu **"mới"** (kiểm định) và đường **màu xanh** biểu thị lỗi trên dữ liệu huấn luyện. Khi chúng ta nhìn vào hình ảnh này, dường như nó đã trả lời cho câu hỏi về **độ phức tạp tối ưu**. Cuối cùng, chúng ta muốn mô hình của mình hoạt động tốt nhất có thể trên dữ liệu mới chứ không phải trên dữ liệu huấn luyện. Có vẻ như việc chọn mức độ phức tạp mà tại đó đường **màu đỏ là thấp nhất** là tối ưu. 

# Thiết lập Kiểm định Chéo như thế nào?
Mặc dù về mặt lý thuyết, quy trình CV rất đơn giản, nhưng rất dễ mắc phải những sai lầm mà trong một số trường hợp sẽ rất tốn kém.

## Phân tầng (Stratify)
Một trong những vấn đề chúng ta có thể gặp phải là sự cân bằng của các nhãn trong tập huấn luyện và kiểm định. Điều này đặc biệt quan trọng khi chúng ta làm việc với **dữ liệu mất cân bằng (imbalanced data)**. Trong các tập dữ liệu có ít nhãn "1" nói chung, rất dễ hình dung một tình huống mà chúng ta rút ra không có hoặc tương đối rất ít nhãn "1". Nó có thể dẫn đến thiên vị và/hoặc phương sai lớn của chỉ số đánh giá trên tập dữ liệu kiểm định. Vì vấn đề này, chúng ta thường buộc phải cân bằng bằng cách sử dụng phân tầng.

Ví dụ, trong trường hợp 5-fold CV, bạn có thể hình dung rằng chúng ta thực hiện việc chia tách tập huấn luyện và kiểm định riêng cho nhãn "1" và "0". Điều này đảm bảo sự cân bằng đồng đều trong cả hai nhóm.

## Phân tích Chuỗi thời gian (Time series analysis)
Tập dữ liệu khó thiết lập kiểm định chéo phù hợp nhất là phân tích chuỗi thời gian. Do thực tế là tất cả các quan sát thường được liên kết (không độc lập theo thời gian), mỗi quan sát phụ thuộc vào các giá trị trước đó. Vì đặc tính này, chúng ta không thể chỉ đơn giản là rút ngẫu nhiên các quan sát vào hai mẫu và sử dụng các hoán vị khác nhau. Các quan sát liền kề nhau về mặt thời gian chứa quá nhiều thông tin về nhau để có thể đặt chúng vào các tập khác nhau (huấn luyện và kiểm định).

Một giải pháp là thực hiện việc chia tách tự nhiên dựa trên thời gian. Chúng ta sử dụng các quan sát **"quá khứ"** làm dữ liệu huấn luyện và các quan sát **"tương lai"** làm tập kiểm định. Giống như trong hình minh họa dưới đây.
<img src="img/cvTimeSeries.png" width="50%">

Nguồn: https://machinelearningmastery.com/backtest-machine-learning-models-time-series-forecasting/

## Rò rỉ Thông tin (Information leakage)
Sự **độc lập** của các quan sát giữa tập huấn luyện và kiểm định là một yêu cầu quan trọng để quy trình mang lại kết quả mong đợi. Tầm quan trọng của nó được thể hiện rõ ràng trong phân tích chuỗi thời gian. Tuy nhiên, rò rỉ thông tin cũng có thể xảy ra ngay cả trong các bài toán phân loại và hồi quy thông thường, khi yêu cầu độc lập không được đáp ứng.

Trong trường hợp này, khi các quan sát không độc lập, chúng ta không thể coi tập kiểm định là dữ liệu *ngoài mẫu* thực sự. Điều này dẫn đến việc **đánh giá thấp lỗi** trên tập kiểm định và có thể dẫn đến việc **đánh giá quá cao** mô hình và kết quả là một mô hình đơn giản là không hoạt động khi triển khai.

Việc có nhiều quan sát cho một người trong bộ dữ liệu là một ví dụ về cấu trúc dữ liệu mà trong đó rất dễ làm mất tính độc lập của dữ liệu. Ví dụ khác có thể là việc sử dụng dữ liệu từ nhiều người trong cùng một hộ gia đình.

Những lỗi trong việc thiết lập quy trình kiểm định chéo rất dễ mắc phải và khó phát hiện. Ngay cả trong các cuộc thi trên Kaggle, rò rỉ dữ liệu vẫn xảy ra theo thời gian mặc dù các tập dữ liệu được chuẩn bị bởi các chuyên gia có kinh nghiệm. Luôn cố gắng tìm hiểu dữ liệu của bạn và thiết lập CV cẩn thận.

# Huấn luyện - Kiểm định - Kiểm tra (Train Validate Test)
Tại thời điểm này, chúng ta nên chỉ ra một vấn đề quan trọng với các tuyên bố được đưa ra ở phần đầu. Chúng ta đã nói rằng chúng ta có thể sử dụng kết quả từ tập kiểm định để chọn giá trị cho các siêu tham số của thuật toán. Tuy nhiên, điều này có nghĩa là chúng ta đang chuyển thông tin từ tập kiểm định sang huấn luyện (rò rỉ dữ liệu - data leakage). Việc này, nếu bị lạm dụng, có thể dẫn đến đánh giá thiên vị và mô hình không tối ưu. Thông thường, rò rỉ thông tin là nhỏ và không đáng kể, tuy nhiên trong những trường hợp chúng ta phụ thuộc nhiều vào tập kiểm định để điều chỉnh thuật toán, chúng ta nên sử dụng quy trình huấn luyện, kiểm định, kiểm tra.

Ý tưởng hoàn toàn giống như kiểm định chéo nhưng chúng ta thêm một bước nữa. Chúng ta giữ lại một phần dữ liệu (test) ở phần đầu mà chúng ta không sử dụng chút nào trong quá trình huấn luyện. Sau đó, cuối cùng, khi mọi thứ đã sẵn sàng, chúng ta kiểm tra mô hình của mình bằng tập kiểm tra này.
<img src="img/cvTVT.png">
Nguồn: https://am207.github.io/2017/wiki/validation.html

## Huấn luyện - Kiểm định - Kiểm tra trong thực tế
Việc chia tập dữ liệu của chúng ta thành ba phần, giống như mọi thứ trong học máy, đều có cả ưu điểm và nhược điểm. Rõ ràng, lợi thế chính là có được đánh giá thuần túy và đáng tin cậy về mô hình. Chúng ta cũng không thể quên những nhược điểm:
* Việc tách tập kiểm tra làm giảm số lượng quan sát trong tập huấn luyện của chúng ta. Nó làm tăng nguy cơ quá khớp và có thể dẫn đến khả năng dự đoán kém hơn của mô hình.
* Trong cách tiếp cận cơ bản, chúng ta chỉ tách một phần của tập dữ liệu làm tập kiểm tra. Điều này có nghĩa là chúng ta lại dựa vào một lần thực hiện của một quá trình ngẫu nhiên nào đó. Điều này có nghĩa là chúng ta có thể nhận được kết quả rất khác nhau mỗi lần ngay cả khi chúng ta lặp lại toàn bộ quy trình huấn luyện giống hệt nhau.

Các vấn đề nói trên có hai giải pháp khả thi. Chúng ta có thể thực hiện k-fold huấn luyện-kiểm tra/kiểm định. Nó làm giảm phương sai của kết quả trong tập kiểm tra. Mặt khác, nó làm tăng độ phức tạp của quy trình ML của chúng ta và không làm giảm vấn đề giảm số lượng quan sát để huấn luyện.

Ngoài ra, chúng ta có thể nói rằng tập kiểm tra thực sự của chúng ta sẽ đến sau. Chúng ta sẽ chỉ cần đợi dữ liệu mới từ thị trường sẽ được tạo ra vào thời điểm mô hình của chúng ta được tạo. Loại dữ liệu kiểm tra này là hoàn hảo vì nó đến từ tương lai. Đây thường là một lựa chọn tốt vì k-fold CV lặp lại được triển khai chính xác sẽ đủ để chọn mô hình tốt nhất và tối ưu hóa các siêu tham số của nó.

Cuối cùng, chúng ta cần chọn một giải pháp phù hợp cho một vấn đề cụ thể, ghi nhớ loại thuật toán chúng ta đang lên kế hoạch sử dụng và hiểu nhu cầu kinh doanh đối với vấn đề ML hiện tại.

## Thiết lập Kiểm định Chéo như thế nào?
Mặc dù về mặt lý thuyết, quy trình CV rất đơn giản, nhưng rất dễ mắc phải những sai lầm mà trong một số trường hợp sẽ rất tốn kém.

### Phân tầng (Stratify)
Một trong những vấn đề chúng ta có thể gặp phải là sự **cân bằng của các nhãn** trong tập huấn luyện và kiểm định. Điều này đặc biệt quan trọng khi chúng ta làm việc với **dữ liệu mất cân bằng (imbalanced data)**.

* Trong các tập dữ liệu có ít nhãn "1" nói chung, rất dễ hình dung một tình huống mà chúng ta rút ra không có hoặc tương đối rất ít nhãn "1".
* Nó có thể dẫn đến **thiên vị** và/hoặc **phương sai lớn** của chỉ số đánh giá trên tập dữ liệu kiểm định.
* Vì vấn đề này, chúng ta thường buộc phải cân bằng bằng cách sử dụng **phân tầng (stratification)**.

**Ví dụ:** Trong trường hợp 5-fold CV, bạn có thể hình dung rằng chúng ta thực hiện việc chia tách tập huấn luyện và kiểm định riêng cho nhãn "1" và "0". Điều này đảm bảo **sự cân bằng đồng đều** trong cả hai nhóm.

### Phân tích Chuỗi thời gian (Time series analysis)
Bộ dữ liệu khó thiết lập kiểm định chéo phù hợp nhất là **phân tích chuỗi thời gian**.
* Do thực tế là tất cả các quan sát thường được **liên kết** (không độc lập theo thời gian), mỗi quan sát phụ thuộc vào các giá trị trước đó.
* Vì đặc tính này, chúng ta không thể chỉ đơn giản là rút ngẫu nhiên các quan sát vào hai mẫu và sử dụng các hoán vị khác nhau.
* Các quan sát liền kề nhau về mặt thời gian chứa quá nhiều thông tin về nhau để có thể đặt chúng vào các tập khác nhau (huấn luyện và kiểm định).

Một giải pháp là thực hiện việc **chia tách tự nhiên dựa trên thời gian**.
* Chúng ta sử dụng các quan sát **"quá khứ"** làm dữ liệu huấn luyện (training data) và các quan sát **"tương lai"** làm tập kiểm định (validation set).
* Giống như trong hình minh họa dưới đây.
<img src="img/cvTimeSeries.png" width="50%">

Nguồn: https://machinelearningmastery.com/backtest-machine-learning-models-time-series-forecasting/

### Rò rỉ Thông tin (Information leakage)
Sự **độc lập** của các quan sát giữa tập huấn luyện (training) và tập kiểm định (validation) là một yêu cầu quan trọng để quy trình mang lại kết quả mong đợi. Tầm quan trọng của nó được thể hiện rõ ràng trong phân tích chuỗi thời gian. Tuy nhiên, **rò rỉ thông tin** vẫn có thể xảy ra ngay cả trong các bài toán phân loại và hồi quy thông thường, khi yêu cầu độc lập không được đáp ứng.

Trong trường hợp này, khi các quan sát không độc lập, chúng ta không thể coi tập kiểm định là dữ liệu *ngoài mẫu* (out of sample) thực sự.
* Điều này dẫn đến việc **đánh giá thấp lỗi** trên tập kiểm định.
* Nó có thể dẫn đến việc **đánh giá quá cao** mô hình và kết quả là một mô hình đơn giản là không hoạt động khi triển khai.

**Ví dụ về cấu trúc dữ liệu dễ gây rò rỉ:**
* Có nhiều quan sát cho cùng một người trong bộ dữ liệu.
* Sử dụng dữ liệu từ nhiều người trong cùng một hộ gia đình.

Các lỗi trong việc thiết lập quy trình kiểm định chéo rất **dễ mắc phải và khó phát hiện**. Ngay cả trong các cuộc thi trên Kaggle, rò rỉ dữ liệu vẫn xảy ra theo thời gian mặc dù các tập dữ liệu được chuẩn bị bởi các chuyên gia có kinh nghiệm. **Luôn cố gắng tìm hiểu dữ liệu của bạn và thiết lập CV cẩn thận**.

## Huấn luyện - Kiểm định - Kiểm tra (Train Validate Test)
Tại thời điểm này, chúng ta nên chỉ ra một vấn đề quan trọng với các tuyên bố được đưa ra ở phần đầu. Chúng ta đã nói rằng chúng ta có thể sử dụng kết quả từ tập kiểm định để chọn giá trị cho các siêu tham số của thuật toán. Tuy nhiên, điều này có nghĩa là chúng ta đang chuyển thông tin từ tập kiểm định sang huấn luyện (**rò rỉ dữ liệu - data leakage**). Việc này, nếu bị lạm dụng, có thể dẫn đến đánh giá thiên vị và mô hình không tối ưu.

Thông thường, rò rỉ thông tin là nhỏ và không đáng kể, tuy nhiên trong những trường hợp chúng ta phụ thuộc nhiều vào tập kiểm định để điều chỉnh thuật toán, chúng ta nên sử dụng quy trình **huấn luyện, kiểm định, kiểm tra** (train, validate, test). Ý tưởng hoàn toàn giống như kiểm định chéo nhưng chúng ta thêm một bước nữa. Chúng ta giữ lại một phần dữ liệu (test) ở phần đầu mà chúng ta không sử dụng chút nào trong quá trình huấn luyện. Sau đó, cuối cùng, khi mọi thứ đã sẵn sàng, chúng ta kiểm tra mô hình của mình bằng tập kiểm tra này.
<img src="img/cvTVT.png">
Source:https://am207.github.io/2017/wiki/validation.html

### Huấn luyện - Kiểm định - Kiểm tra trong thực tế
Việc chia tập dữ liệu của chúng ta thành ba phần, giống như mọi thứ trong học máy, đều có cả ưu điểm và nhược điểm. Rõ ràng, lợi thế chính là có được **đánh giá thuần túy và đáng tin cậy** về mô hình. Chúng ta cũng không thể quên những nhược điểm:
* Việc tách tập kiểm tra làm giảm số lượng quan sát trong tập huấn luyện của chúng ta. Nó làm tăng nguy cơ quá khớp và có thể dẫn đến khả năng dự đoán kém hơn của mô hình.
* Trong cách tiếp cận cơ bản, chúng ta chỉ tách một phần của tập dữ liệu làm tập kiểm tra. Điều này có nghĩa là chúng ta lại dựa vào một lần thực hiện của một quá trình ngẫu nhiên nào đó. Điều này có nghĩa là chúng ta có thể nhận được kết quả rất khác nhau mỗi lần ngay cả khi chúng ta lặp lại toàn bộ quy trình huấn luyện giống hệt nhau.

Các vấn đề nói trên có hai giải pháp khả thi.
1.  Chúng ta có thể thực hiện **k-fold huấn luyện-kiểm tra/kiểm định** (train-test/validate). Nó làm giảm phương sai của kết quả trong tập kiểm tra. Mặt khác, nó làm tăng độ phức tạp của quy trình ML của chúng ta và không làm giảm vấn đề giảm số lượng quan sát để huấn luyện.
2.  Ngoài ra, chúng ta có thể nói rằng tập kiểm tra thực sự của chúng ta sẽ đến sau. Chúng ta sẽ chỉ cần đợi **dữ liệu mới từ thị trường** sẽ được tạo ra vào thời điểm mô hình của chúng ta được tạo. Loại dữ liệu kiểm tra này là hoàn hảo vì nó đến từ tương lai. Đây thường là một lựa chọn tốt vì k-fold CV lặp lại được triển khai chính xác sẽ đủ để chọn mô hình tốt nhất và tối ưu hóa các siêu tham số của nó.

Cuối cùng, chúng ta cần chọn một giải pháp phù hợp cho một vấn đề cụ thể, ghi nhớ loại thuật toán chúng ta đang lên kế hoạch sử dụng và hiểu nhu cầu kinh doanh đối với vấn đề ML hiện tại.

## Cross validation in Python
Most of the methods for cross validation are really well implemented in sklearn.

http://scikit-learn.org/stable/modules/cross_validation.html


In [ ]:
import pandas as pd
import numpy as np
import pickle
import statsmodels.api as sm
import matplotlib.pyplot as plt
import gc
from sklearn.metrics import roc_auc_score

plt.style.use('seaborn-ticks')
%matplotlib inline

Lets get back to our medical set.

In [ ]:
%cd C:
medical = pd.read_csv("medical_care.csv")
print(medical.shape)
pd.set_option("display.max_columns", 50)
medical.head()

C:\programy\Python39\lib\site-packages\scipy\.libs
(35072, 29)


UMARSTAT UCUREMP UCURNINS              USATMED  URELATE  \
0          Never married      No      Yes       Very satisfied        2   
1              Separated     Yes       No       Very satisfied        2   
2  Married_live together      No       No       Very satisfied        5   
3               Divorced      No      Yes  Little dissatisfied        4   
4          Never married     Yes       No       Very satisfied        0   

    REGION STATE      HHID FHOSP  FDENT  FEMER  FDOCT         UIMMSTAT  \
0  Midwest    WI  55616128    No      0      0      0  US-born citizen   
1  Midwest    WI  54704000    No      2      0      0  US-born citizen   
2  Midwest    WI  57874272    No      0      1      0  US-born citizen   
3  Midwest    WI  54106816    No      0      0      1  US-born citizen   
4  Midwest    WI  54569152    No      2      0      0  US-born citizen   

  U_USBORN  UAGE     U_FTPT  U_WKSLY  U_HRSLY  U_USHRS  HEARNVAL  HOTHVAL  \
0      Yes    22  Full-time     52.0       40       40         0        0   
1      Yes    30  Full-time     52.0       40       40     31468     5950   
2      Yes    33  Part-time     52.0       30       30     24700    11340   
3      Yes    41  Part-time     43.0       40       25     60000    39002   
4      Yes    34  Full-time     52.0       40       40     55280     4200   

   HRETVAL  HSSVAL  HWSVAL UBRACE  GENDER  \
0        0       0       0  White  Female   
1        0       0   31468  White  Female   
2        0    4920   24700  White    Male   
3        0       0   60000  Black  Female   
4        0       0   55280  Black    Male   

                                    UEDUC3  CEYES  CHAIR  
0                     No HS diploma or GED  hazel  brown  
1  HS diploma or GED, no bachelor's degree   blue  black  
2                     No HS diploma or GED  brown  brown  
3                     No HS diploma or GED  brown  black  
4  HS diploma or GED, no bachelor's degree  brown  black

Now lets encode UCURNINS variable and run the regression on our full dataset.

In [ ]:
medical["UCURNINS"] = (medical.UCURNINS == "Yes").astype(int)

In [ ]:
mod = sm.GLM.from_formula(formula = 'UCURNINS ~ UMARSTAT + USATMED + URELATE + REGION + FHOSP + FDENT + FEMER + FDOCT + ' + 
                          'UIMMSTAT + UAGE + U_FTPT + U_WKSLY + U_USHRS + HOTHVAL + HRETVAL + HSSVAL + HWSVAL + UBRACE + ' + 
                          'UEDUC3 + GENDER',
                          data = medical,
                          family = sm.families.Binomial())
res = mod.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:               UCURNINS   No. Observations:                35072
Model:                            GLM   Df Residuals:                    35036
Model Family:                Binomial   Df Model:                           35
Link Function:                  Logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -11126.
Date:                Tue, 29 Mar 2022   Deviance:                       22251.
Time:                        19:37:05   Pearson chi2:                 4.05e+04
No. Iterations:                     6   Pseudo R-squ. (CS):             0.1742
Covariance Type:            nonrobust                                         
=====================================================================================================================
                                                        coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------------------------------
Intercept                                             0.9019      0.216      4.166      0.000       0.478       1.326
UMARSTAT[T.Married, do not live together]            -0.4264      0.166     -2.563      0.010      -0.752      -0.100
UMARSTAT[T.Married_live together]                    -0.8331      0.056    -14.774      0.000      -0.944      -0.723
UMARSTAT[T.Never married]                            -0.3337      0.064     -5.210      0.000      -0.459      -0.208
UMARSTAT[T.Partnership]                               0.3966      0.085      4.677      0.000       0.230       0.563
UMARSTAT[T.Separated]                                -0.0527      0.095     -0.553      0.580      -0.239       0.134
UMARSTAT[T.Unknown]                                   0.6720      0.388      1.732      0.083      -0.089       1.433
UMARSTAT[T.Widowed]                                  -0.1727      0.144     -1.197      0.231      -0.455       0.110
USATMED[T.Little satisfied]                          -0.4597      0.063     -7.324      0.000      -0.583      -0.337
USATMED[T.No opinion]                                 0.6913      0.098      7.025      0.000       0.498       0.884
USATMED[T.Very dissatisfied]                          0.5547      0.092      6.017      0.000       0.374       0.735
USATMED[T.Very satisfied]                            -0.7287      0.062    -11.668      0.000      -0.851      -0.606
REGION[T.Northeast]                                  -0.0486      0.057     -0.858      0.391      -0.159       0.062
REGION[T.South]                                       0.7221      0.047     15.439      0.000       0.630       0.814
REGION[T.West]                                        0.4020      0.049      8.143      0.000       0.305       0.499
FHOSP[T.Yes]                                         -0.2438      0.075     -3.254      0.001      -0.391      -0.097
UIMMSTAT[T.Foreign-born, non-citizen]                 0.7271      0.089      8.172      0.000       0.553       0.901
UIMMSTAT[T.US-born citizen]                          -0.5706      0.080     -7.114      0.000      -0.728      -0.413
U_FTPT[T.Part-time]                                   0.4966      0.058      8.526      0.000       0.382       0.611
UBRACE[T.Asian/Pacific Islander]                     -0.9828      0.161     -6.104      0.000      -1.298      -0.667
UBRACE[T.Black]                                      -0.4685      0.126     -3.710      0.000      -0.716      -0.221
UBRACE[T.White]                                      -0.5119      0.118     -4.338      0.000      -0.743      -0.281
UEDUC3[T.HS diploma or GED, no bachelor's degree]     0.8375      0.053     15.828      0.000       0.734       0.941
UEDUC3[T.No HS 

In [ ]:
roc_auc_score(y_test, preds)

NameError: name 'y_test' is not defined

Đối với các lần chia đơn giản, chúng ta có thể sử dụng **train_test_split**. Phương thức này sẽ chia tập dữ liệu thành các biến đầu vào (features) và cột biến phụ thuộc (dependant variable) của chúng ta.
* Chúng ta có thể định nghĩa **random_state** hoặc khởi tạo nó bằng một số ngẫu nhiên.
* Việc kiểm soát tính ngẫu nhiên của các lần chia có thể giúp ích rất nhiều cho khả năng **tái lập kết quả** trong một số trường hợp.
* Để thấy sự khác biệt, hãy chạy lại một vài lần ô phía trên và phía dưới.

In [ ]:
from sklearn.model_selection import train_test_split
import random

X_train, X_test, y_train, y_test = train_test_split(medical,
                                                    medical.UCURNINS,
                                                    test_size = 0.3,
                                                    random_state = random.randint(0, 1000))
print(X_train.shape, X_test.shape)
mod = sm.GLM.from_formula(formula = 'UCURNINS ~ UMARSTAT + USATMED + URELATE + REGION + FHOSP + FDENT + FEMER + FDOCT + ' + 
                          'UIMMSTAT + UAGE + U_FTPT + U_WKSLY + U_USHRS + HOTHVAL + HRETVAL + HSSVAL + HWSVAL + UBRACE + ' + 
                          'UEDUC3 + GENDER',
                          data = X_train,
                          family = sm.families.Binomial())
res = mod.fit()
res.summary()
preds = res.predict(X_test)

roc_auc_score(y_test, preds)

Chúng ta có thể rất dễ dàng áp dụng **phân tầng (stratify)** chỉ với một đối số. Chúng ta chỉ cần đặt cột cho việc phân tầng.

In [ ]:
scores = []

for k in range(100) :
    
    X_train, X_test, y_train, y_test = train_test_split(medical,
                                                        medical.UCURNINS,
                                                        stratify = medical.UCURNINS,
                                                        test_size = 0.3,
                                                        random_state = random.randint(0, 10000))
    # print(X_train.shape, X_test.shape)
    mod = sm.GLM.from_formula(formula = 'UCURNINS ~ UMARSTAT + USATMED + URELATE + REGION + FHOSP + FDENT + FEMER + FDOCT + ' + 
                              'UIMMSTAT + UAGE + U_FTPT + U_WKSLY + U_USHRS + HOTHVAL + HRETVAL + HSSVAL + HWSVAL + UBRACE + ' + 
                              'UEDUC3 + GENDER',
                              data = X_train,
                              family = sm.families.Binomial())
    res = mod.fit()
    res.summary()
    preds = res.predict(X_test)
    scores.append(roc_auc_score(y_test, preds))
    # print(scores[-1])

Như chúng ta có thể thấy, trong trường hợp hồi quy logistic trên bộ dữ liệu này, có **rất ít sự khác biệt** khi có và không có phân tầng. Nhãn "1" của chúng ta được đại diện đủ tốt (**5 000 quan sát**) để tính ngẫu nhiên hoạt động đủ tốt với tỷ lệ chia 0.7 và 0.3.

In [ ]:
medical.UCURNINS.value_counts()

#### Let's see some diagnostics...

In [ ]:
df = pd.DataFrame(data = scores, columns = ['scores'])
print(df.scores.describe())
print(f'Kurtosis \t %s' %round(df.scores.kurtosis(), 5)) # This one returns 'normalized' kurtosis (i.e. 0 instead of 3).
print(f'Skewness \t %s' %round(df.scores.skew(), 5))

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

ax = sns.displot(data = df.scores, kde = True, label = 'empirical hist')
x0, x1 = ax.ax.get_xlim()
x_pdf = np.linspace(x0, x1, len(df))
y_pdf = stats.norm.pdf(x_pdf, df.scores.mean(), df.scores.std())
ax.ax.plot(x_pdf, y_pdf, 'r', lw = 2, label = 'normal pdf')
ax.ax.legend()

In [ ]:
from scipy import stats 

alpha = 1e-3

k2, p = stats.normaltest(df.scores) # Kurtosis and skewness normality test.

if p < alpha :

    print('Kurtosis, Skewness test: The null hypothesis about normality can be rejected.')

else:

    print('Kurtosis, Skewness test: The null hypothesis about normality can not be rejected.')

# Kolmogorov Smirnov normality test.
ks = stats.kstest(df.scores, 'norm')

if ks[1] < alpha :

    print('Kolmogorov Smirnov test: The null hypothesis about normality can be rejected.')

else:

    print('Kolmogorov Smirnov test: The null hypothesis about normality can not be rejected.')

# Do you remember that @kstest function requires NORMALIZATION of sample?
ks = stats.kstest((df.scores - df.scores.mean()) / df.scores.std(), 'norm')

if ks[1] < alpha :

    print('Kolmogorov Smirnov (normalized sample) test: The null hypothesis about normality can be rejected.')

else:

    print('Kolmogorov Smirnov (normalized sample) test: The null hypothesis about normality can not be rejected.')

Bây giờ chúng ta có thể kiểm tra **xu hướng quá khớp (overfitting)** trong hồi quy logistic.
* Như bạn có thể thấy bên dưới, chúng ta cần giảm kích thước tập dữ liệu huấn luyện xuống chỉ còn **10%** để thực sự bắt đầu quá khớp.
* Điều này cho thấy khả năng **chống lại vấn đề quá khớp** là rất tốt.

In [ ]:
for k in range(1, 10) :
    
    X_train, X_test, y_train, y_test = train_test_split(medical,
                                                        medical.UCURNINS,
                                                        test_size = 0.1 * k,
                                                        random_state = 0)
    mod = sm.GLM.from_formula(formula = 'UCURNINS ~ UMARSTAT + USATMED + URELATE + REGION + FHOSP + FDENT + FEMER + FDOCT + ' + 
                              'UIMMSTAT + UAGE + U_FTPT + U_WKSLY + U_USHRS + HOTHVAL + HRETVAL + HSSVAL + HWSVAL + UBRACE + ' + 
                              'UEDUC3 + GENDER',
                              data = X_train,
                              family = sm.families.Binomial())
    res = mod.fit()
    predsTrain = res.predict(X_train)
    preds = res.predict(X_test)
    print("Train AUC:", round(roc_auc_score(y_train, predsTrain), 4), "Valid AUC:", round(roc_auc_score(y_test, preds), 4))

Đối với k-fold CV, chúng ta không nhận được tập dữ liệu mà là các **chỉ mục (indices)** cho các tập dữ liệu. Điều này được thực hiện để tránh **trùng lặp tập dữ liệu không cần thiết**.

Hãy chạy 10-fold CV và xem kết quả. Đối với k-fold, chúng ta cần thêm **shuffle = True**.

In [ ]:
from sklearn.model_selection import KFold

kf = KFold(n_splits = 10, shuffle = True, random_state = random.randint(0, 10000))

for train, test in kf.split(medical.index.values) :
    
    mod = sm.GLM.from_formula(formula = 'UCURNINS ~ UMARSTAT + USATMED + URELATE + REGION + FHOSP + FDENT + FEMER + FDOCT + ' + 
                              'UIMMSTAT + UAGE + U_FTPT + U_WKSLY + U_USHRS + HOTHVAL + HRETVAL + HSSVAL + HWSVAL + UBRACE + ' + 
                              'UEDUC3 + GENDER',
                              data = medical.iloc[train],
                              family = sm.families.Binomial())
    res = mod.fit()
    predsTrain = res.predict(medical.iloc[train])
    preds = res.predict(medical.iloc[test])
    print("Train AUC:", round(roc_auc_score(medical.iloc[train].UCURNINS, predsTrain), 4), "Valid AUC:",
          round(roc_auc_score(medical.iloc[test].UCURNINS, preds), 4))


Như chúng ta có thể thấy, kết quả **thay đổi rất nhiều từ mẫu này sang mẫu khác**. Chúng ta thấy các giá trị từ 0.81 đến 0.838 trên các tập kiểm định.

Để hiểu rõ hơn về vấn đề **phương sai**, hãy xem kết quả cho 5-fold CV.
* Phương sai của kết quả dường như **thấp hơn một chút** đối với 10-fold CV.
* Điều này là tự nhiên vì hồi quy logistic không quá khớp nhiều và chúng ta **tăng kích thước của tập kiểm định** (điều này ổn định kết quả trên các mẫu kiểm định).

In [ ]:
kf = KFold(n_splits = 5, shuffle = True, random_state = random.randint(0, 10000))

for train, test in kf.split(medical.index.values) :
    
    mod = sm.GLM.from_formula(formula = 'UCURNINS ~ UMARSTAT + USATMED + URELATE + REGION + FHOSP + FDENT + FEMER + FDOCT + ' + 
                              'UIMMSTAT + UAGE + U_FTPT + U_WKSLY + U_USHRS + HOTHVAL + HRETVAL + HSSVAL + HWSVAL + UBRACE + ' + 
                              'UEDUC3 + GENDER',
                              data = medical.iloc[train],
                              family = sm.families.Binomial())    
    res = mod.fit()
    predsTrain = res.predict(medical.iloc[train])
    preds = res.predict(medical.iloc[test])
    print("Train AUC:", round(roc_auc_score(medical.iloc[train].UCURNINS, predsTrain), 4),
          "Valid AUC:", round(roc_auc_score(medical.iloc[test].UCURNINS, preds), 4))


Hãy xem vấn đề **phương sai** chi tiết hơn. Hãy chạy **10-fold CV 10 lần** và xem phương sai của các giá trị trung bình.

In [ ]:
for z in range(10) :
    
    trainRes = []
    valRes = []
    kf = KFold(n_splits = 10, shuffle = True, random_state = random.randint(0, 10000))
    
    for train, test in kf.split(medical.index.values):
        
        mod = sm.GLM.from_formula(formula = 'UCURNINS ~ UMARSTAT + USATMED + URELATE + REGION + FHOSP + FDENT + FEMER + FDOCT + ' + 
                              'UIMMSTAT + UAGE + U_FTPT + U_WKSLY + U_USHRS + HOTHVAL + HRETVAL + HSSVAL + HWSVAL + UBRACE + ' + 
                              'UEDUC3 + GENDER',
                              data = medical.iloc[train],
                              family = sm.families.Binomial())          
        res = mod.fit()
        predsTrain = res.predict(medical.iloc[train])
        preds = res.predict(medical.iloc[test])
        trainRes.append(roc_auc_score(medical.iloc[train].UCURNINS, predsTrain))
        valRes.append(roc_auc_score(medical.iloc[test].UCURNINS, preds))
        
    print("Train AUC:", round(np.mean(trainRes), 4), "Valid AUC:", round(np.mean(valRes), 4))


Now let's do the same for 5-fold CV.

In [ ]:
for z in range(10) :
    
    trainRes = []
    valRes = []
    kf = KFold(n_splits = 5, shuffle = True, random_state = random.randint(0, 10000))
    
    for train, test in kf.split(medical.index.values) :
        
        mod = sm.GLM.from_formula(formula = 'UCURNINS ~ UMARSTAT + USATMED + URELATE + REGION + FHOSP + FDENT + FEMER + FDOCT + ' + 
                              'UIMMSTAT + UAGE + U_FTPT + U_WKSLY + U_USHRS + HOTHVAL + HRETVAL + HSSVAL + HWSVAL + UBRACE + ' + 
                              'UEDUC3 + GENDER',
                              data = medical.iloc[train],
                              family = sm.families.Binomial()) 
        res = mod.fit()
        predsTrain = res.predict(medical.iloc[train])
        preds = res.predict(medical.iloc[test])
        trainRes.append(roc_auc_score(medical.iloc[train].UCURNINS, predsTrain))
        valRes.append(roc_auc_score(medical.iloc[test].UCURNINS, preds))
        
    print("Train AUC:", round(np.mean(trainRes), 4), "Valid AUC:", round(np.mean(valRes), 4))

Có vẻ như đối với hồi quy logistic, kết quả **rất ổn định** cho cả kiểm định 5-fold và 10-fold. Hơn nữa, nó **không quá khớp nhiều** (overfit), vì sự khác biệt về giá trị của chỉ số đo lường (metric) cho tập huấn luyện và tập kiểm định là **tối thiểu**.

In [ ]:
predList = []
indList = []
kf = KFold(n_splits = 5, shuffle = True, random_state = random.randint(0, 10000))

for train, test in kf.split(medical.index.values) :
    
    mod = sm.GLM.from_formula(formula = 'UCURNINS ~ UMARSTAT + USATMED + URELATE + REGION + FHOSP + FDENT + FEMER + FDOCT + ' + 
                          'UIMMSTAT + UAGE + U_FTPT + U_WKSLY + U_USHRS + HOTHVAL + HRETVAL + HSSVAL + HWSVAL + UBRACE + ' + 
                          'UEDUC3 + GENDER',
                          data = medical.iloc[train],
                          family = sm.families.Binomial()) 
    res = mod.fit()
    predsTrain = res.predict(medical.iloc[train])
    preds = res.predict(medical.iloc[test])
    
    predList.append(preds.tolist())
    indList.append(medical.iloc[test].index.tolist())
    
    trainRes.append(roc_auc_score(medical.iloc[train].UCURNINS, predsTrain))
    valRes.append(roc_auc_score(medical.iloc[test].UCURNINS, preds))
    
print("Train AUC:", round(np.mean(trainRes), 4), "Valid AUC:", round(np.mean(valRes), 4))

In [ ]:
predsSorted = pd.Series(sum(predList, []), index = sum(indList, [])).sort_index()
roc_auc_score(medical.UCURNINS.sort_index(), pd.Series(sum(predList, []), index = sum(indList, [])).sort_index())

In [ ]:
from sklearn.model_selection import StratifiedKFold

predList = []
indList = []
kf = StratifiedKFold(n_splits = 5, shuffle = True, random_state = random.randint(0, 10000))

for train, test in kf.split(medical.index.values, medical.UCURNINS) :
    
    mod = sm.GLM.from_formula(formula = 'UCURNINS ~ UMARSTAT + USATMED + URELATE + REGION + FHOSP + FDENT + FEMER + FDOCT + ' + 
                          'UIMMSTAT + UAGE + U_FTPT + U_WKSLY + U_USHRS + HOTHVAL + HRETVAL + HSSVAL + HWSVAL + UBRACE + ' + 
                          'UEDUC3 + GENDER',
                          data = medical.iloc[train],
                          family = sm.families.Binomial())
    res = mod.fit()
    predsTrain = res.predict(medical.iloc[train])
    preds = res.predict(medical.iloc[test])
    
    predList.append(preds.tolist())
    indList.append(medical.iloc[test].index.tolist())
    
    trainRes.append(roc_auc_score(medical.iloc[train].UCURNINS, predsTrain))
    valRes.append(roc_auc_score(medical.iloc[test].UCURNINS, preds))
    
print("Train AUC:", round(np.mean(trainRes), 4), "Valid AUC:", round(np.mean(valRes), 4))

#### Nhưng, đừng quên về chất lượng mô hình tổng thể...

Phần lớn mọi người đều có bảo hiểm, chỉ bằng cách **đoán bừa** rằng ai đó có bảo hiểm, chúng ta đã đúng trong **85%** trường hợp, và chất lượng tổng thể của mô hình của chúng ta chỉ cao hơn một chút so với con số này.

In [ ]:
print(f'Overall percent of uninsured persons is %s' %round(medical.UCURNINS.sum() / medical.UCURNINS.count() * 100, 2))

In [ ]:
print(f'Quality score for training set is %s' 
      %round(sum(medical.UCURNINS.iloc[train] == (predsTrain > 0.5) * 1) / len(train) * 100, 2))
print(f'Quality score for test set is %s' 
      %round(sum(medical.UCURNINS.iloc[test] == (preds > 0.5) * 1) / len(test) * 100, 2))

#--------------------------------------------------------------------
## Bài tập 2.

### Bài tập 2.1.

**Dữ liệu hành khách Titanic – 1310 quan sát và 15 biến:**
* **passenger_id** – Id hành khách duy nhất
* **pclass** – Hạng vé (1 = hạng 1, 2 = hạng 2, 3 = hạng 3)
* **survived** – Sống sót (0 = Không, 1 = Có)
* **name** – Tên và Họ
* **sex** – Giới tính (0 = Nam, 1 = Nữ)
* **age** – Tuổi tính bằng năm
* **sibsp** – Số anh chị em ruột / vợ chồng trên tàu Titanic
* **parch** – Số cha mẹ / con cái trên tàu Titanic
* **ticket** – Số vé
* **fare** – Giá vé hành khách
* **cabin** – Số cabin
* **embarked** – Cảng lên tàu (C = Cherbourg, Q = Queenstown, S = Southampton)
* **boat** – Thuyền cứu sinh (nếu sống sót)
* **body** – Số thi thể (nếu không sống sót và thi thể được tìm thấy)
* **home.dest** – Quê quán/Điểm đến

---

**Yêu cầu:**

1.  **Chạy lại các mô hình tốt nhất** của bạn cho tất cả các thuật toán với **5-fold CV**.
2.  **Kiểm tra sự ổn định** của kết quả cho **K-fold lặp lại**.
3.  Kiểm tra trong k-fold CV lặp lại xem việc thêm **phân tầng** có thay đổi kết quả của bạn không (sự ổn định).
4.  Kiểm tra xem bạn có bị **quá khớp** trong các mô hình của mình không. Kiểm tra xem bạn có thể cải thiện điểm kiểm định (**validation score**) của mình không.


### Bài tập 2.2.
**Bộ Dữ liệu Chất lượng Rượu: "data/wines.csv"**
* Nguồn: https://archive.ics.uci.edu/ml/datasets/wine+quality
* Tệp chứa dữ liệu về các mẫu rượu vang trắng và đỏ Bồ Đào Nha **Vinho Verde**.
* Các đặc điểm lý hóa khác nhau của từng mẫu có sẵn cũng như **điểm chất lượng rượu** trên thang điểm (0-10) được thực hiện bởi các chuyên gia.

---

**Yêu cầu:**

1.  **Chạy lại các mô hình tốt nhất** của bạn cho tất cả các thuật toán với **5-fold CV**.
2.  **Kiểm tra sự ổn định** của kết quả cho **K-fold lặp lại**.
3.  Kiểm tra trong k-fold CV lặp lại xem việc thêm **phân tầng** có thay đổi kết quả của bạn không (sự ổn định).
4.  **So sánh ảnh hưởng của phân tầng** với vấn đề Titanic.
5.  Kiểm tra xem bạn có bị **quá khớp** trong các mô hình của mình không. Kiểm tra xem bạn có thể cải thiện điểm kiểm định (**validation score**) của mình không.